### ISSUE TITLE
Problem initiating a transcode on a slave backend.

### ISSUE THREAD

**Issue by tekrantz** [MythTV-Clients/MythtvPlayerForAndroid#3]
Apologies in advance.  I do not yet have a stack trace for this.  I will get one.

When I try to queue an HLS transcode of a recording made on a slave backend I get a failure indicating that some java component timed out trying to reach the slave backend after resolving the name of the machine to an IP address.  The IP address that was retrieved was incorrect because it is all internal to my network and I do not have a complete DNS solution for my internal boxes. The transcode never starts and eventually mythtvPlayer crashed.

So, why am I posting about the problem here?  Because I used the mythweb:6544 interface in a web browser and selected the same recording for streaming.  The transcode started and I was able to view the recording in JWPlayer.

Also interestingly, after mythweb:6544 started the transcode I was able to select that recording in myththPlayer and it "knew" that a transcoded stream was available and was able to play it.

Obviously the DNS issues on my internal network are NOT your issue but the fact that mythweb:6544 was able to queue the transcode makes me wonder if mythtvPlayer could do that also?

**Comment by billmeek** [comment]
If it helps:

I noted that if the app's setup is changed to the slave BE's IP, transcoding works
and the recording can be watched. If the master's IP is restored, then the following fires:
<pre>
02-13 13:33:37.170  28703-28780/org.mythtv.android D/Retrofit﹕ ---- ERROR http://192.168.1.204:6544/Content/GetLiveStreamList [.204 is my master BE]
02-13 13:33:37.170  28703-28780/org.mythtv.android D/Retrofit﹕ java.net.UnknownHostException: Unable to resolve host "slave-mc1": No address associated with hostname
            at java.net.InetAddress.lookupHostByName(InetAddress.java:400)
            at java.net.InetAddress.getAllByNameImpl(InetAddress.java:242)
            at java.net.InetAddress.getAllByName(InetAddress.java:220)
            at com.squareup.okhttp.internal.Dns$1.getAllByName(Dns.java:29)
            at com.squareup.okhttp.internal.http.RouteSelector.resetNextInetSocketAddress(RouteSelector.java:231)
            at com.squareup.okhttp.internal.http.RouteSelector.next(RouteSelector.java:124)
            at com.squareup.okhttp.internal.http.HttpEngine.connect(HttpEngine.java:317)
            at com.squareup.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:241)
            at com.squareup.okhttp.Call.getResponse(Call.java:198)
            at com.squareup.okhttp.Call.execute(Call.java:80)
            at retrofit.client.OkClient.execute(OkClient.java:53)
            at org.mythtv.services.api.EtagInterceptingOkClient.execute(EtagInterceptingOkClient.java:70)
            at retrofit.RestAdapter$RestHandler.invokeRequest(RestAdapter.java:326)
            at retrofit.RestAdapter$RestHandler.invoke(RestAdapter.java:240)
            at $Proxy3.getLiveStreamList(Native Method)
            at org.mythtv.android.library.core.service.v028.content.ContentServiceV28EventHandler.getLiveStreamInfoList(ContentServiceV28EventHandler.java:63)
            at org.mythtv.android.library.core.MainApplication$RefreshLiveStreamsTask.doInBackground(MainApplication.java:258)
            at org.mythtv.android.library.core.MainApplication$RefreshLiveStreamsTask.doInBackground(MainApplication.java:250)
            at android.os.AsyncTask$2.call(AsyncTask.java:264)
            at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:305)
            at java.util.concurrent.FutureTask.run(FutureTask.java:137)
            at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:208)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1076)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:569)
            at java.lang.Thread.run(Thread.java:856)
02-13 13:33:37.170  28703-28780/org.mythtv.android D/Retrofit﹕ ---- END ERROR
02-13 13:33:37.170  28703-28780/org.mythtv.android E/ContentServiceV28EventHandler﹕ getLiveStreamInfoList : error - Unable to resolve host "slave-mc1": No address associated with hostname
</pre>

When Play is pressed, the screen goes black and after a 3 minutes the "Cannot play video"
popup displays. The screen goes blank again.

Seems like the question becomes: should the app (or some other configuration elsewhere) 
contain a list of slave BEs (without rooting the device.)

**Comment by dmfrey** [comment]
I will take a look and see if the recording details provides any insight
into which host it resides on.

I feel this is a fundamental issue with how the master backend and
secondary backends are setup.  IMO, we should only be communicating with
the master backend.  It should be responsible for triggering jobs such as
this on the other backends that are on the network.

**Comment by billmeek** [comment]
In http://mc0:6544/Dvr/GetRecordedList, Program->HostName is
the host that did the original recording. Once transcoded, it's in
http://mc0:6544/Content/GetLiveStreamList, LiveStreamInfo->SourceHost

**Comment by dmfrey** [comment]
I don't have anyway to map a hostname to connected backend.  For instance, i am not running bind, so I can't just use my master backend hostname as a connection url.

I was looking through the /Myth endpoint to see if there was a way that I could determine it. It looks like it is only dealing in hostnames, from what I can tell.  I am not sure I can assume a user is running bind.

What is the expectation over a ssh connection if you are not a home?

**Comment by tekrantz** [comment]
Not sure if this is helpful or not but using firefox on my phone on a free wifi connection I was able to schedule a transcode of a recording that was made on a slave backend over the an ssh port forward connection.  I made no attempt to play it over that connection but I am guessing that if I tried with mythPlayer I could have played it.

Basically I used an SSH connection (with connectBot) from my phone to my home firewall.  Then I had an SSH port forward of phone port 1234 to my backend IP address port 6544.  I then pointed firefox on my phone to 127.0.0.1:1234 to get at the mythweb. Then I selected on recording from the slave backend to be transcoded.

**Comment by dmfrey** [comment]
@tekrantz was this ssh tunnel to your master backend or secondary backend?
I am guessing you port forward 6544 to your master backend and 1234 to your
secondary backend.

I think its a broader question...how best to handle multiple backends.
Does the services api give us a "map" of the myth environment?  Right now,
I don't think so.  Also, accessing multiple backends, remotely would
require a user to open multiple ssh tunnels and open multiple ports in the
firewall.  Maybe a vpn would be a better solution than multiple ssh tunnels
(just thinking out loud).

**Comment by tekrantz** [comment]
@tekrantz was this ssh tunnel to your master backend or secondary backend?
I am guessing you port forward 6544 to your master backend and 1234 to your
secondary backend.

The tunnel was to my master backend.  I had NO tunnel to my slave backend. Port 1234 on my phone was forwarded to 6544 on my master backend.  That was the only tunnel I had going.  But somehow mythweb was still able to begin the transcode with no connection to the slave backend.  

 

Please understand that I am not pressing this issue just trying to give information.

**Comment by billmeek** [comment]
@dmfrey, the MAF Bonjour method worked for getting the IP and hostname.
Although there were some (at least 1) user that didn't want to run avahi on
their hosts.

Adding to the complexity of settings, the hostname and IP address could
be stored and then MPFA would do the lookup (conversion from host to
IP) internally. There will no doubt be those that have old recordings that
have a previous hostname (we know some users change host names 
during the life of their MythTV use.)

I'd suggest that there's some value is continuing to discuss this because
it was one of the items raised in MAF.

Although I completely agree that the 1st choice would be to do everything
through the frontend. I don't see anything available in the Services API
either.

**Comment by dmfrey** [comment]
@tekrants can you verify that when performing a HLS transcode through
mythweb, it took place on the secondary backend, not the master backend? I
am wondering if there is a discrepancy between how the services request a
live stream transcode from the backend and how the mythweb is doing it.

- Please understand that I am not pressing this issue just trying to give
information.
I understand.  Please don't hold back with the questions or issues you are
seeing. It is only going to help to make the app better.

@billmeek I agree that bonjour would help in this case, however, I feel we
got a lot of blowback for requiring that.  I also agree that this is a good
conversation. Maybe we need to rope in @cpinkham to see what it would take
for the services to only have to communicate with the master backend
alone.  I don't think the app should be required to talk to multiple
backends.

**Comment by billmeek** [comment]
It looks like the host issue is in the HTTP header returned by MythTV, not in the
MythTV Services API (at this point.) We're talking to the master BE and it's
redirecting to the slave. That seem to be a good thing.
<pre>
 curl --include mc0:6544/Content/AddRecordingLiveStream?RecordedId=2352&ChanId=1261&StartTime=2015-02-14T20%3A00%3A00.000Z&MaxSegments=0&Width=1280&Height=720
 HTTP/1.1 301 Unknown [Wikipedia says 301 = Moved Permanently]
 Accept-Ranges: bytes
 Connection: Keep-Alive
 Content-Length: 0
 Content-Type: 
 Date: 14 Feb 2015 23:51:02
 Keep-Alive: timeout=10
 Location: http://slave-mc1:6544/Content/AddRecordingLiveStream?RecordedId=2352
</pre>
The problem remains, Android doesn't have an easily changed /etc/hosts.

**Comment by dmfrey** [comment]
It looks like the OkHttp client used to back Retrofit calls to the backend
needs to be configured to follow redirects.  I think this should solve the
issue.  I am going to make a change to the services api and push it.  Then
we can check this once its propagated to the maven repositories.

**Comment by dmfrey** [comment]
I have this implemented locally.  I will push it shortly. Still waiting on maven central to synchronize.  However, one thought comes to mind.  If I am on an ssh tunnel to the master backend,  I am uncertain how the app is going to be able to follow a 301 permanent redirect over that tunnel.  So I do not feel this will ultimately solve the issue.  But I do feel enabling follow redirects in the api is a good thing, so I am going to leave this in.

**Comment by billmeek** [comment]
Short answer, still fails.  Better info: my *curl* above was for the wrong endpoint.
But the result is the same. The 1st issue (and stacktrace) is from this:
<pre>
curl --include 'mc0:6544/Content/GetPreviewImage?ChanId=1091&StartTime=2015-02-13T18:03:00'
HTTP/1.1 301 Unknown
...
Location: http://slave-mc1:6544/Content/GetPreviewImage?ChanId=1091&StartTime=2015-02-13T18:03:00
</pre>

The above commit doesn't fix the error because the hostname is returned
in the redirect and Android isn't able to resolve it. This only prevents the
preview from being seen. 

The more  important portion of the problem is that the request to actually 
play the file is sent to the master BE, where it fails because there's no HLS
there. A 404 is returned.
<pre>
curl -i 192.168.1.204:6544/StorageGroup/Streaming/1261_20150214200000.ts.1280x720_800kV_64kA.m3u
HTTP/1.1 404 Not Found
</pre>
The same logic used by GetPreviewImage could probably be
added to MythTV, but unless an IP address is returned, the
result would be the same. Plus I'm assuming the initial 
mythtranscode can be done on the slave, which isn't true
right now. Can 'o worms.

**Comment by billmeek** [comment]
Turns out the the app uses: Content/AddRecordingLiveStream and that uses
the same hostname check as GetPreviewImage. And it too redirects to the
proper (slave in this case) host. Still stuck  at the lack of name resolution.
Now, if the backend only redirected to an IP address...

**Comment by dmfrey** [comment]
@billmeek there are a few add live stream commands. One for adding
recordings, one for adding videos, and a general one.  Maybe the last one
is the one we should be using.  I don't recall what we were using in the
old app, but I don't think it is very different.

**Comment by billmeek** [comment]
Both AddVideoLiveStream and AddRecordingLiveStream do the
host check and redirect if required. Then they both call
AddLiveStream. I'd say you're using the proper calls. I didn't
look at the MAF code, but that's a good idea. I'm trying to write
a standalone Qt program to test the QHostInfo::lookupHost()
because it can return the IP address. If it works, it would be up
to the MythTV developers to agree that redirecting to an IP is
better than a hostname.

**Comment by billmeek** [comment]
@tekrantz, if you're building MythTV from source, consider trying 
this patch (cd yourSource/mythtv/mythtv;git apply theFollowingInSomeFile):
<pre>
diff --git a/mythtv/programs/mythbackend/services/content.cpp b/mythtv/programs/mythbackend/services/content.cpp
index e25affc..2e1fe81 100644
--- a/mythtv/programs/mythbackend/services/content.cpp
+++ b/mythtv/programs/mythbackend/services/content.cpp
@@ -1044,7 +1044,10 @@ DTC::LiveStreamInfo *Content::AddRecordingLiveStream(
 
         LOG(VB_UPNP, LOG_ERR, sMsg);
 
-        throw HttpRedirectException( pginfo.GetHostname() );
+        //throw HttpRedirectException( pginfo.GetHostname() );
+        QString addrstr = gCoreContext->resolveAddress(
+            pginfo.GetHostname(), gCoreContext->ResolveIPv4, false);
+        throw HttpRedirectException( addrstr );
     }
 
     QString sFileName( GetPlaybackURL(&pginfo) );
</pre>
It returns an IPv4 address, as long as your slave backend is defined
in /etc/hosts. I should try it running bind, IPv6, error handling...
<pre>
$ curl --include 'mc0:6544/Content/AddRecordingLiveStream?RecordedId=2352'
...
Location: http://192.168.1.223:6544/Content/AddRecordingLiveStream?RecordedId=2352
</pre>

**Comment by billmeek** [comment]
@dmfrey, the patch for MythTV (above) has been rewritten and works,
*if* I hardcode my slave backend IP:6544 here:

https://github.com/MythTV-Clients/MythtvPlayerForAndroid/blob/master/mobile/src/main/java/org/mythtv/android/player/PlayerActivity.java#L184/blob/master/mobile/src/main/java/org/mythtv/android/player/PlayerActivity.java#L184

rather that use the master:6544 plus the relative URL. The full URL would work
(except for IPv6 users like me), but that's another issue. I can't figure out how to
get the full URL. The relative URL is developed on line 110.

The big question is how much further to go. Your point about remote users and
port forwarding seems to be the remaining issue. I've made MythTV do the host
name resolution but there's no sense submitting the patch if it's of no help for
users of this app. Please let me know what you'd like to do.

**Comment by billmeek** [comment]
Grrr, just deleted all the experiments and started a fresh xcode from the
app with the above patch. It's xcoding correctly on the slave, however, the IP
address returned in FullURL from Content/GetLiveStreamList is that
of my master BE!

**Comment by dmfrey** [comment]
@billmeek I think what would be nice here is if the backend would reverse proxy the call.  This way, we would only be communicating with the master backend.  The master backend would proxy all the calls to the respective frontends. 

Maybe we should write up what the current behavior is and the desired behavior should be.  Maybe then we could come up with a solution.  Can you add what you have found so far to a wiki page on what mythtv currently does?  Then, maybe we can point one of the devs to it for confirmation/explanation. Then we can tackle the issue on our end.

**Comment by billmeek** [comment]
@dmfrey, OK, now that I've read about 'reverse proxy', that sounds like
a perfect solution. I don't see anything else using it in MythTV or in the
Qt website, so I don't think I can write a patch to do it. That leaves the
new MythTV Forum's feature request section (Trac tickets without a
patch are closed immediately.) I'm able to do a write-up there.

But 1st:

The actual player sends an IP address/Storage Group/Filename, and
I don't understand how that would work with reverse proxy. How do
you see that working with reverse proxy?
```
    url=http://192.168.1.204:6544/StorageGroup/Streaming/1261_20150214200000.ts.1280x720_800kV_64kA.m3u8
```
Currently the BE puts the Master IP in the FullUrl, for slaves too
 but that's a 2 line fix. I've already tried it.

I was also looking at the iPad version of this app that was written by
a developer who's left the project. Interesting notes here:

    http://trianythingonce.com/blog/torc-for-ios/ 
    https://groups.google.com/forum/#!topic/torc-for-ios/SvQwkZI7pxA

The iPad, like android doesn't have a hosts equivalent unless it's
jailbroken.

**Comment by dmfrey** [comment]
@tekrantz Can you test this again.  I just pushed a change for Issue #37. I think it might fix this as well.

**Comment by tekrantz** [comment]
This can be closed.  I just grabbed the latest compiled it.  I selected a recording that was recorded on a slave backend and it began transcoding on the master backend and began to play perfectly.

 

@tekrantz <https://github.com/tekrantz>  Can you test this again. I just pushed a change for Issue #37 <https://github.com/MythTV-Clients/MythtvPlayerForAndroid/issues/37> . I think it might fix this as well.

**Comment by dmfrey** [comment]
@tekrants thank you

@billmeek I will make my change the permanent solution for this and Issue
#37

**Comment by billmeek** [comment]
Just finished testing on my 0.28-pre slave and got the same results as
@tekrantz. Note that there is at least one MythTV user that prefers
to transcode on a more powerful slave. I believe he was a MAF beta
user, but I'm not going to look it up (nor reopen this issue.) If it comes
up, then it can be a new issue and will require an upstream fix too.
AddLiveStream uses myth:// protocol, while AddRecordingLiveStream
does a check for the hostname then either calls AddLiveStream or
sends the redirect response.
